src/ngraph_assign_clusters.cc

```diff
@@ -559,12 +653,88 @@ Status AssignClusters(Graph* graph) {
 
       // TODO(amprocte): move attr name to a constant
       node->AddAttr("_ngraph_cluster", cluster_idx);
+
+      if (config::IsLoggingPlacement()) {
+        // map from cluster id to ngraph_cluster id
+        cluster_to_encapsulate[cluster->index] = cluster_idx;
+      }
     }
 
     seen.insert(cluster);
   }
   NGRAPH_VLOG(2) << "Tagging done";
 
+  if (config::IsLoggingPlacement()) {
+    std::cout << "\n=============Edge contraction logs=============\n";
+    int num_reasons = 7;  // the number of elements in the reasons enum
+    vector<int> reason_count(num_reasons,
+                             0);  // histogram of reasons of non-contraction
+    int num_non_contracted = 0;
+    std::vector<string> reason_string(  // to convert the enum to string
+        {"NOTANOP", "UNSUPPORTED", "DEADNESS", "BACKEND", "SAMECLUSTER",
+         "STATICINPUT", "PATHEXISTS"});
+    std::cout
+        << "_ngraph_cluster i->j: non contraction reason histogram (Cannot be "
+           "UNSUPPORTED or NOTANOP because unsupported ops will not be "
+           "assigned an "
+           "encapsulate)\n";
+    for (auto it : cluster_separation_reason) {
+      num_non_contracted += it.second.size();
+      vector<string> reasons_for_this_edge;
+      for (auto& inner_itr : it.second) {
+        reason_count[inner_itr]++;
+        reasons_for_this_edge.push_back(reason_string[inner_itr]);
```

Comment:
Regarding the data structure named, reasons_for_this_edge.
From what i understand, there can be only one (known) reason for an edge to not contract, as we exit from the loop as soon as we get a reason. There can be more than one reasons for two clusters to not merge (one reason for each edge between the clusters). If this is correct, then please consider renaming it.

Reply:
yes it should be `reasons_for_this_separation` as clusters can have multiple reasons to separate, but edges will have only 1 reason (well, they may have 2, but there is an ordering)